AZ3166/src/libraries/WebSocket/src/WebSocketClient.cpp

```diff
@@ -0,0 +1,448 @@
+{
+    int ret = parseURL(url, scheme, sizeof(scheme), host, sizeof(host), &port, path, sizeof(path));
+    if (ret)
+    {
+        Serial.println("ERROR: URL parsing failed; please use: \"ws://ip-or-domain[:port]/path\".");
+        return;
+    }
+
+    if (port == 0) //TODO: do handle WSS->443
+    {
+        port = 80;
+    }
+
+    if (strcmp(scheme, "ws"))
+    {
+        Serial.println("ERROR: Wrong scheme, please use \"ws\" instead.");
+    }
+}
+
+int WebSocketClient::parseURL(const char *url, char *scheme, size_t maxSchemeLen, char *host, size_t maxHostLen, uint16_t *port, char *path, size_t maxPathLen)
+{
+    char *schemePtr = (char *)url;
+    char *hostPtr = (char *)strstr(url, "://");
+    if (hostPtr == NULL)
+    {
+        Serial.println("ERROR: Could not find host from the WebSocket URL.");
+        return -1;
+    }
+
+    if (maxSchemeLen < hostPtr - schemePtr + 1) // Including NULL-terminating char
+    {
+        Serial.printf("ERROR: Scheme str is too small (%d >= %d)", maxSchemeLen, hostPtr - schemePtr + 1);
+        return -1;
+    }
+    memcpy(scheme, schemePtr, hostPtr - schemePtr);
+    scheme[hostPtr - schemePtr] = '\0';
+
+    hostPtr += 3;
+    size_t hostLen = 0;
+
+    char *portPtr = strchr(hostPtr, ':');
+    if (portPtr != NULL)
+    {
+        hostLen = portPtr - hostPtr;
+        portPtr++;
+        if (sscanf(portPtr, "%hu", port) != 1)
+        {
+            Serial.println("ERROR: Could not find port.");
+            return -1;
+        }
+    }
+    else
+    {
+        *port = 0;
+    }
+    char *pathPtr = strchr(hostPtr, '/');
+    if (pathPtr == NULL)
+    {
+        Serial.println("ERROR: Path not specified. Please add /[path] to the end of the websocket address");
+        return -1;
+    }
+    if (hostLen == 0)
+    {
+        hostLen = pathPtr - hostPtr;
+    }
+
+    if (maxHostLen < hostLen + 1) //including NULL-terminating char
+    {
+        Serial.printf("ERROR: Host str is too small (%d >= %d)", maxHostLen, hostLen + 1);
+        return -1;
+    }
+    memcpy(host, hostPtr, hostLen);
+    host[hostLen] = '\0';
+
+    size_t pathLen;
+    char *fragmentPtr = strchr(hostPtr, '#');
+    if (fragmentPtr != NULL)
+    {
+        pathLen = fragmentPtr - pathPtr;
+    }
+    else
+    {
+        pathLen = strlen(pathPtr);
+    }
+
+    if (maxPathLen < pathLen + 1) //including NULL-terminating char
+    {
+        Serial.printf("ERROR: Path str is too small (%d >= %d)", maxPathLen, pathLen + 1);
+        return -1;
+    }
+    memcpy(path, pathPtr, pathLen);
+    path[pathLen] = '\0';
+
+    return 0;
+}
+
+bool WebSocketClient::connect()
+{
+    if (_tcpSocket != NULL)
+    {
+        // Already connected.
+        return true;
+    }
+    else
+    {
+        _tcpSocket = new TCPSocket();
+        if (_tcpSocket == NULL)
+        {
+            return false;
+        }
+
+        if (_tcpSocket->open(WiFiInterface()) != 0 || _tcpSocket->connect(host, port) != 0)

```

Comment:
consider not hide the WiFiInterface() here, let caller pass in the NetworkInterface that can make this library more generic

Reply:
Since we ONLY have one NetworkInterface in our DevKit SDK, keep it temporaly

---
In reply to: [173728218](https://github.com/Microsoft/devkit-sdk/pull/771#discussion_r173728218) [](ancestors = 173728218)